### PR TITLE
Add Beam pipeline to generate *srvpb.FileDecorations

### PR DESCRIPTION
**Pull request by schroederc**
`write_tables --experimental_beam_pipeline` now generates a LevelDB that
can be used to serve file decorations.  The decorations are nearly
complete.  The missing target definitions, overrides, and diagnostics
will be added in future commits.

### CODE REVIEW

File: kythe/go/serving/pipeline/beam.go [hunk 1]
```diff
@@ -58,6 +75,162 @@ func FromEntries(s beam.Scope, entries beam.PCollection) *KytheBeam {
...
+
+func fileVName(anchor *spb.VName) *spb.VName {
+	return &spb.VName{
+		Corpus: anchor.Corpus,
+		Root:   anchor.Root,
+		Path:   anchor.Path,
+	}
+}
+
+func nodeToDecorPiece(key *spb.VName, node func(**ppb.Node) bool, file func(**spb.VName) bool, emit func(*spb.VName, *ppb.DecorationPiece)) {
```
Comment (jrtom) [inline kythe/go/serving/pipeline/beam.go]: Can you help me understand how Beam provides the *ToDecorPiece functions with their inputs?  (I assume you're doing so correctly, but I don't understand the mechanism at all.)

File: kythe/go/serving/pipeline/beam.go [hunk 2]
```diff
@@ -76,33 +248,45 @@ func (k *KytheBeam) References() beam.PCollection {
...
+}
+
 func keyByPath(n *ppb.Node) (*spb.VName, *ppb.Node) {
 	return &spb.VName{Corpus: n.Source.Corpus, Root: n.Source.Root, Path: n.Source.Path}, n
 }
 
-func toRefs(p *spb.VName, file func(**ppb.Node) bool, anchor func(**ppb.Node) bool, emit func(*ppb.Reference)) error {
-	var fileNode *ppb.Node
-	if !file(&fileNode) {
+func toRefs(p *spb.VName, file func(**srvpb.File) bool, anchor func(**ppb.Node) bool, emit func(*ppb.Reference)) error {
```
Comment (jrtom) [inline kythe/go/serving/pipeline/beam.go]: Please explain this change (and possibly update the PR description).  The PR description doesn't make it sound as though there's any bugs being fixed, and this seems pretty clearly a semantic change (i.e., not just new capabilities being added).

Comment (schroederc) [reply kythe/go/serving/pipeline/beam.go]: This isn't fixing anything.  I changed the input from taking a raw `*ppb.Node` representing a file to take a `*srvpb.File`.  I did this to share the `*ppb.Node -> *srvpb.File` transform for the file decorations.

File: kythe/go/serving/pipeline/beam.go [hunk 3]
```diff
@@ -58,6 +75,162 @@ func FromEntries(s beam.Scope, entries beam.PCollection) *KytheBeam {
...
+func (c *combineDecorPieces) CreateAccumulator() *srvpb.FileDecorations {
+	return &srvpb.FileDecorations{}
+}
+
+func (c *combineDecorPieces) MergeAccumulators(accum, n *srvpb.FileDecorations) *srvpb.FileDecorations {
+	return accum
+}
+
+func (c *combineDecorPieces) AddInput(accum *srvpb.FileDecorations, p *ppb.DecorationPiece) *srvpb.FileDecorations {
+	if ref := p.GetReference(); ref != nil {
```
[the remark this answers is not in the document]
Comment (schroederc) [reply kythe/go/serving/pipeline/beam.go]: ✓

File: kythe/go/serving/pipeline/beam.go [hunk 4]
```diff
@@ -58,6 +75,162 @@ func FromEntries(s beam.Scope, entries beam.PCollection) *KytheBeam {
...
+func toEnclosingFile(r *ppb.Reference) (*spb.VName, *spb.VName, error) {
+	anchor, err := kytheuri.ToVName(r.Anchor.Ticket)
+	if err != nil {
+		return nil, nil, err
+	}
+	file := fileVName(anchor)
+	return r.Source, file, nil
+}
+
+// combineDecorPieces :: *ppb.DecorationPiece -> *srvpb.FileDecorations
```
[the remark this answers is not in the document]
Comment (schroederc) [reply kythe/go/serving/pipeline/beam.go]: Sorry, I meant that to be "is typed".

File: kythe/go/serving/pipeline/beam.go (hunk 1 above)
Comment (schroederc) [reply kythe/go/serving/pipeline/beam.go]: I'll take `nodeToDecorPiece` as an example.

`nodeToDecorPiece` is a ParDo function called on line 89 and takes a CoGBK as its input.  CoGBKs can be thought as a set of tuples with the first component being a key and then each subsequent component being a collection of values (a GBK is essentially just a CoGBK of 1 collection of values per key).  Beam represents the collections as streams/iterators (i.e. that's `func(*ppb.Node) bool` and `func(**spb.VName) bool`).  The key is just the `*spb.VName`.

The CoGBK values come from the input tables to the CoGBK on line 90: the PCollection of `KV<*spb.VName, *ppb.Node>` and the PColleciton of `KV<*spb.VName, *srvpb.File>`.  Each table is grouped by key and its value collection per key becomes the stream passed to the subsequent ParDo.  Each tuple is passed independently to `nodeToDecorPiece` to be processed.

**Pull request merged**